src/server/pfs/server/driver.go

```diff
@@ -2186,25 +2265,77 @@ func (d *driver) getFile(ctx context.Context, file *pfs.File, offset int64, size
 	if err != nil {
 		return nil, err
 	}
-
 	paths, err := tree.Glob(file.Path)
 	if err != nil {
 		return nil, err
 	}
+	var objects []*pfs.Object
+	var totalSize int64
+	var sortedPaths []string
+	foundDirectoryNode := false
 	for path, node := range paths {
-		if node.FileNode == nil {
-			delete(paths, path)
+		sortedPaths = append(sortedPaths, path)
+		if node.DirNode != nil {
+			foundDirectoryNode = true
 		}
 	}
+	if !foundDirectoryNode {
+		// e.g. this was a request for a single file e.g. /foo/bar.txt
+		dirNodePaths, err := tree.Glob(filepath.Dir(file.Path))
+		if err != nil {
+			return nil, err
+		}
+		if len(dirNodePaths) != 1 {
+			return nil, fmt.Errorf("invalid set of directory node paths - found %v expected 1 - this is probably a bug", len(dirNodePaths))
+		}
+		for key, node := range dirNodePaths {
+			paths[key] = node
+			sortedPaths = append(sortedPaths, key)
+		}
+	}
+	sort.Strings(sortedPaths)
+	footers := stack.New()
+	directories := stack.New()
+	for _, path := range sortedPaths {
+		node := paths[path]
+		thisDir := directories.Peek()
+		if thisDir != nil && !strings.HasPrefix(path, thisDir.(string)) {
+			// We've proceeded past the current directory
+			footer := footers.Pop().(*pfs.Object)
+			if footer != nil {
```

Comment:
Probably better to do this check before you push rather than after you pop, for deeply nested directory structures pushing and popping for every one could actually start to be a meaningful amount of memory moving around.

Reply:
So, as-is, the code handles the case where we traverse a directory w a header and no footer. If I only add the footer if its non nil, then this check needs to change as well. We'd still need to check if its non nil to make sure we handle the case where the stack is empty (otherwise we'd panic when traversing a directory only w a header). But it would save some space on the stack. So I'll go ahead and add this test case and make these changes.